text/0000-cargo-feature-migrations.md

```diff
@@ -0,0 +1,388 @@
+- Feature Name: `cargo-feature-migrations`
+- Start Date: 2021-07-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Extend Cargo to allow some limited forms of migrations of feature sets to be expressed.
+This will allow new features to be added that make existing functionality optional without causing needless breakage.
+
+# Motivation
+[motivation]: #motivation
+
+## Problem
+
+Today, Cargo's features most easily support a workflow where features added in new versions of a crate gate new functionality.
+
+> Example: In a new version of a crate, someone creates a new function, and adds it behind a new feature.
+> Neither the function nor feature existed in the old version of the crate.
+
+The problem is there is another quite different use case of features: making previously mandatory functionality optional.
+
+> Example: In a new version of a create, some creates a new feature for a function that *already* existed.
+> Before, the export of that function was mandatory, but now it is optional because it can be disabled.
+
+The workflow isn't supported so well: all Cargo offers for it is the notion of "default features", which isn't sufficient for reasons that will be elaborated below.
+
+This second use-case is really important — in fact, perhaps more important.
+The key thing aspect about features to realize is that the ultimate *reason* they are used is not to control a crate's interface (exports), but its *dependencies* (imports).
+All things equal, no one minds if a crate provides some extra functionality they don't need.
```

Comment:
All things are not equal; I care about build time.

Reply:
I changed to "a little". Please bear with me :) in now way does this proposal cause regressions to the first case, so it shouldn't cause an increase in build times. I just merely want to point out my opinion that the support for the latter use-case is too small given its importance.